sdk/dotnet/Pulumi/Core/Input.cs

```diff
@@ -42,10 +43,37 @@ private protected Input(Output<T> outputValue)
         public static implicit operator Output<T>(Input<T> input)
             => input._outputValue;
 
-        public Output<T> ToOutput()
-            => this;
-
         IOutput IInput.ToOutput()
-            => ToOutput();
+            => this.ToOutput();
+    }
+
+    public static class InputExtensions
+    {
+        /// <summary>
+        /// <see cref="Output{T}.Apply{U}(Func{T, Output{U}?})"/> for more details.
+        /// </summary>
+        public static Output<U> Apply<T, U>(this Input<T>? input, Func<T, U> func)
+            => input.ToOutput().Apply(func) ?? Output.Create(default(U)!);
+
+        /// <summary>
+        /// <see cref="Output{T}.Apply{U}(Func{T, Output{U}?})"/> for more details.
+        /// </summary>
+        public static Output<U> Apply<T, U>(this Input<T>? input, Func<T, Task<U>> func)
+            => input.ToOutput().Apply(func) ?? Output.Create(default(U)!);
+
+        /// <summary>
+        /// <see cref="Output{T}.Apply{U}(Func{T, Output{U}?})"/> for more details.
+        /// </summary>
+        public static Output<U> Apply<T, U>(this Input<T>? input, Func<T, Input<U>?> func)
+            => input.ToOutput().Apply(func) ?? Output.Create(default(U)!);
+
+        /// <summary>
+        /// <see cref="Output{T}.Apply{U}(Func{T, Output{U}?})"/> for more details.
+        /// </summary>
+        public static Output<U> Apply<T, U>(this Input<T>? input, Func<T, Output<U>?> func)
+            => input.ToOutput().Apply(func) ?? Output.Create(default(U)!);
+
+        public static Output<T> ToOutput<T>(this Input<T>? input)
+            => input ?? Output.Create(default(T)!);
```

Comment:
`.ToOutput` can be called on a null Input to always get back to a non-null output.